pkg/cli/destroy/destroy.go

```diff
@@ -240,38 +256,101 @@ func Destroy(ctx context.Context, opts DeleteOptions) error {
-		list, err := resourceClient.List(ctx, metav1.ListOptions{})
-		if err != nil {
-			return false, err
-		}
-		if len(list.Items) == 0 {
-			return true, nil
+	//  function to poll with wait.ExponentialBackoffWithContext
+	deleteAndWait := func(deleteFinalizers bool) func(ctx context.Context) (bool, error) {
+		// log each key as waiting only once on the info levell, and continue logging on the debug level
+		loggedDeletion := sets.New[string]()
+		infofOnceThenDebugf := func(str string, args ...interface{}) {
+			logLine := fmt.Sprintf(str, args...)
+			if loggedDeletion.Has(logLine) {
+				log.Debug(logLine)
+				return
+			}
+			log.Info(logLine)
+			loggedDeletion.Insert(logLine)
 		}
-		for _, object := range list.Items {
-			if !object.GetDeletionTimestamp().IsZero() {
-				return false, nil
+
+		return func(ctx context.Context) (bool, error) {
+			infofOnceThenDebugf("checking all %q", resource)
+
+			// fetch all
+			resourceClient := dynamicClient.Resource(gvr)
+			list, err := resourceClient.List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return false, err
 			}
-			if object.GetNamespace() == "" {
-				log.Infof("deleting %v: %v", resource, object.GetName())
-			} else {
-				log.Infof("deleting %v: %v/%v", resource, object.GetNamespace(), object.GetName())
+			// succeed when all resources are deleted
+			if len(list.Items) == 0 {
+				return true, nil
 			}
-			err := resourceClient.Namespace(object.GetNamespace()).Delete(ctx, object.GetName(), metav1.DeleteOptions{})
-			if err != nil && !kerrors.IsNotFound(err) {
-				return false, err
+
+			isVCluster := resource == "virtualclusterinstances"
+
+			// delete all resources and log deleting resources
+			for _, object := range list.Items {
+				// get namespaced name
+				namespacedName := object.GetName()
+				namespace := object.GetNamespace()
+				if namespace != "" {
+					namespacedName += "/" + namespace
+				}
+				isExternalVCluster := false
+				if isVCluster {
+					//convert unstructured to VirtualClusterInstance
+					virtualClusterInstance := &storagev1.VirtualClusterInstance{}
+					err = runtime.DefaultUnstructuredConverter.FromUnstructured(object.Object, &virtualClusterInstance)
+					if err != nil {
+						log.Warnf("couldn't cast %q object %q to VirtualClusterInstance: %v", resource, namespacedName, err)
+					}
+					isExternalVCluster = virtualClusterInstance.Spec.External
+				}
+				// delete object if not already deleted
+				if object.GetDeletionTimestamp().IsZero() {
+					if !isExternalVCluster {
+						log.Infof("deleting %v: %q", resource, namespacedName)
+					} else {
+						log.Infof("deleting externally deployed %v, the virtual cluster itself will remain: %q", resource, namespacedName)
+					}
+					err := resourceClient.Namespace(object.GetNamespace()).Delete(ctx, object.GetName(), metav1.DeleteOptions{})
+					if kerrors.IsNotFound(err) {
+						continue
+					} else if err != nil {
+						return false, err
+					}
```

Comment:
Spoke to Lukas regarding external cluster that still may be irreparably affected by deletion. At the moment this only is cluster that use db connector and it will not be common for new features to have a similar effect. So we can make another exception here and provide a prompt. You can see if a vcluster is using a db connector by checking if there are any secrets with the `loft.sh/database-vcluster` set to `virtualClusterInstance.Status.ServiceUID`. If a vcluster is external _and_ has one of these secrets, provide a prompt with wording similar to "Warning: `<vclustername>` is an external cluster but has a datastore managed by platform. Deleting this will destroy your virtual cluster's datastore and it will no longer be usable as the vcluster pod will enter a bad state."

Reply:
Added the changes!

Reply:
However, I parsed the object's config instead of searching for secrets